ui-episodedetails/src/main/java/app/tivi/episodedetails/EpisodeDetailsViewModel.kt

```diff
@@ -55,29 +56,38 @@ internal class EpisodeDetailsViewModel @Inject constructor(
     private val removeEpisodeWatch: RemoveEpisodeWatch,
     private val logger: Logger,
     private val snackbarManager: SnackbarManager
-) : ReduxViewModel<EpisodeDetailsViewState>(
-    EpisodeDetailsViewState(
-        // The string "episodeId" is the name of the argument in the route
-        episodeId = savedStateHandle.get("episodeId")!!
-    )
-) {
+) : ViewModel() {
+
+    private val episodeId: Long = savedStateHandle.get("episodeId")!!
 
     private val loadingState = ObservableLoadingCounter()
 
     private val pendingActions = MutableSharedFlow<EpisodeDetailsAction>()
 
-    init {
-        viewModelScope.launch {
-            observeEpisodeDetails.observe()
-                .collect { updateFromEpisodeDetails(it) }
-        }
-
-        viewModelScope.launch {
-            observeEpisodeWatches.observe()
-                .onStart { emit(emptyList()) }
-                .collect { updateFromEpisodeWatches(it) }
-        }
+    val state = combine(
+        observeEpisodeDetails.observe(),
+        observeEpisodeWatches.observe(),
+        loadingState.observable,
+        snackbarManager.flow.onStart { emit(null) },
+    ) { episodeDetails, episodeWatches, refreshing, error ->
+        EpisodeDetailsViewState(
+            episodeId,
+            episode = episodeDetails.episode,
+            season = episodeDetails.season,
+            watches = episodeWatches,
+            canAddEpisodeWatch = episodeDetails.episode?.firstAired?.let {
+                it.isBefore(OffsetDateTime.now())
+            } ?: true,
+            refreshing = refreshing,
+            error = error,
+        )
+    }.stateIn(
+        scope = viewModelScope,
+        started = WhileSubscribed(5000),
```

Comment:
This is only helpful if you're collecting the flow using the new `addRepeatingJob` API or you're manually cancelling the job/scope when the view goes to the background. This won't work if you're using `launchWhenStarted`.

Which one are you using?

Reply:
Hmmm, these are collected from Compose via `collectAsState()`. Oh well.

Reply:
Ah! Then, yes, you're having the same problems unless you remember a `Flow.flowWithLifecycle` as you can read here: https://medium.com/androiddevelopers/a-safer-way-to-collect-flows-from-android-uis-23080b1f8bda

Relevant code snippet: https://gist.github.com/manuelvicnt/6629b53e33b97c127508227ed6e0bee4#file-flowwithlifecyclecompose-kt

Reply:
Migrated to `flowWithLifecycle()` 👍